monitoring/ping.go

```diff
@@ -0,0 +1,263 @@
+	// finding what is the current node
+	var self serf.Member
+	for _, node := range nodes {
+		// cut the port portion of the address after the ":" away
+		if node.Name == c.serfMemberName {
+			self = node
+		}
+	}
+	if self.Name == "" {
+		return trace.NotFound("self node Serf Member not found for %s", c.serfMemberName)
+	}
+
+	// ping each other node and fail in case the results are over a specified
+	// threshold
+	for _, node := range nodes {
+		// skip pinging self
+		if self.Addr.String() == node.Addr.String() {
+			continue
+		}
+
+		rttNanoSec, err := calculateRTT(client, self, node)
+		if err != nil {
+			return err
+		}
+
+		err = c.storePingInHDR(rttNanoSec, node)
+		if err != nil {
+			return err
+		}
+
+		roundtripLatencyInterface, _ := c.roundtripLatency.Get(node.Name)
+		roundtripLatency, ok := roundtripLatencyInterface.(*hdrhistogram.Histogram)
+		if !ok {
+			return trace.Errorf("couldn't parse roundtripLatency as HDRHistogram on %s", c.serfMemberName)
+		}
+		log.Debugf("%s <-ping-> %s = %dns [latest]", self.Name, node.Name, rttNanoSec)
+		log.Debugf("%s <-ping-> %s = %dns [%.2f percentile]",
+			self.Name, node.Name,
+			roundtripLatency.ValueAtQuantile(pingRoundtripQuantile),
+			pingRoundtripQuantile)
+
+		pingRoundtripPercentile := roundtripLatency.ValueAtQuantile(pingRoundtripQuantile)
+		if pingRoundtripPercentile >= pingRoundtripThreshold.Nanoseconds() {
+			log.Warningf("%s <-ping-> %s = slow ping RoundTrip detected. Value %dns over threshold %s (%dns)",
+				self.Name, node.Name, pingRoundtripPercentile,
+				pingRoundtripThreshold.String(), pingRoundtripThreshold.Nanoseconds())
+		} else {
+			log.Debugf("%s <-ping-> %s = ping RoundTrip okay. Value %dns within threshold %s (%dns)",
+				self.Name, node.Name, pingRoundtripPercentile,
+				pingRoundtripThreshold.String(), pingRoundtripThreshold.Nanoseconds())
+		}
+	}
+
+	return nil
+}
+
+// storePingInHDR is used to store ping RoundTrip values in HDR Histograms in memory
+func (c *pingChecker) storePingInHDR(pingroundtripLatency int64, node serf.Member) error {
+	s, exists := c.roundtripLatency.Get(node.Name)
+	if !exists {
+		c.roundtripLatency.Set(node.Name,
+			hdrhistogram.New(pingRoundtripMinimum.Nanoseconds(),
+				pingRoundtripMaximum.Nanoseconds(),
+				pingRoundtripSignificativeFigures),
+			statsTTLPeriod)
+		s, _ = c.roundtripLatency.Get(node.Name)
+	}
+
+	nodeLatencies, ok := s.(*hdrhistogram.Histogram)
+	if !ok {
+		return trace.BadParameter("couldn't parse roundtripLatency as HDRHistogram on %s", c.serfMemberName)
+	}
+
+	tmpSnapshot := nodeLatencies.Export()
+	if len(tmpSnapshot.Counts) >= slidingWindowSize {
+		// pop element at index 0 (oldest)
+		nodeLatencies.Reset()
+
+		lowerLimit := len(tmpSnapshot.Counts) - slidingWindowSize
+		if lowerLimit < 0 {
+			lowerLimit = 0
+		}
+		// using len(tmpSnapshot.Counts)-1 as the upper limit as another value
+		// will be added a few lines below
+		for i := lowerLimit; i < len(tmpSnapshot.Counts)-1; i++ {
```

Comment:
Hmm, I don't think this has an effect that you indented it to have. If you look at how the hdrhistogram is implemented, you'll see that it stores values in a "sparse" list ("counts" field you use here) - when you record a value in it, it calculates the value's "index" and then increments the value in the list at that index by 1. So iterating over "counts" here does not actually mean iterating over values oldest-to-newest.

If you want to keep a sliding window of values, I think that instead of storing histogram in the map, you will need to store a list of values as they arrive and then construct histogram on demand in the check. I.e. (semi-pseudo-Go):

```go
type pingChecker struct {
    windows TTLMap // TTL map where key is node name and value is `[]int64`
}

func (c *pingChecker) storePingValue(node, value) error {
  // first record value for the node in the list, throwing out the older element if needed
  latencies, ok := c.windows.Get(node) // latencies is a slice of latencies
  // see if we need to truncate the list
  if len(latencies) >= maxWindowSize {
    latencies = latencies[1:]
  }
  // append new value
  latencies = append(latencies, value)
  // build hdr histogram
  histogram := hdrhistogram.New(...)
  for _, latency := range latencies {
    histogram.RecordValue(latency)
  }
  // check percentiles, etc.
}
```
Let me know if that makes sense.

Reply:
makes sense, I'll reimplement it following your suggestion.